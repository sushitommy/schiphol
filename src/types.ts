export type AppState = {
  inputValue: string;
  data: Flight[] | null;
  isLoading: boolean;
  isError: boolean;
  sorting: SortingType;
};

export interface APIResponse extends Response {
  flights: Flight[];
}

export type Flight = {
  flightIdentifier: string;
  flightNumber: string;
  airport: string;
  date: string;
  expectedTime: string;
  originalTime: string;
  url: string;
  score: number;
};

export enum ActionType {
  SET_DATA_LOADING,
  SET_DATA_ERROR,
  SET_DATA,
  SET_INPUT_VALUE,
  SET_SORTING
}

type SetDataLoadingAction = { type: ActionType.SET_DATA_LOADING; payload: boolean };
type SetDataErrorAction = { type: ActionType.SET_DATA_ERROR; payload: boolean };
type SetData = { type: ActionType.SET_DATA; payload: Flight[] | null };
type SetInputValueAction = { type: ActionType.SET_INPUT_VALUE; payload: string };
type SetSortingValueAction = { type: ActionType.SET_SORTING; payload: SortingType };

export type Action = SetDataLoadingAction | SetDataErrorAction | SetData | SetInputValueAction | SetSortingValueAction;

export enum SortingType {
  EARLY_TO_LATE = "earlyToLate",
  LATE_TO_EARLY = "lateToEarly"
}
