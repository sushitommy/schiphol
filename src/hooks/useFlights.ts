import { useEffect, useState } from "react";
import { APIResponse, Flight } from "types";

export default function useFlightData(url: string = "./flights.json") {
  const [data, setData] = useState<Flight[] | null>(null);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [isError, setError] = useState<boolean>(false);

  useEffect(() => {
    let isMounted = true;

    const fetchData = async () => {
      setIsLoading(true);
      try {
        const response = await fetch(url);
        if (!response.ok) {
          throw new Error(`Error: ${response.status}`);
        }
        const result: APIResponse = await response.json();
        if (isMounted) {
          setData(result.flights);
          setIsLoading(false);
        }
      } catch (error) {
        if (isMounted) {
          setError(true);
          setIsLoading(false);
        }
      }
    };

    fetchData();

    return () => {
      isMounted = false;
    };
  }, [url]);

  return { data, isLoading, isError };
}
