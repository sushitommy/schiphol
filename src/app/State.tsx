import useFlights from "@hooks/useFlights";
import { PropsWithChildren, createContext, useEffect, useReducer } from "react";
import { Action, ActionType, AppState, SortingType } from "types";

const initialState: AppState = {
  inputValue: "",
  inputHasChanged: false,
  data: [],
  isLoading: false,
  isError: false,
  sorting: SortingType.EARLY_TO_LATE
};

function reducer(state: AppState, action: Action) {
  switch (action.type) {
    case ActionType.SET_DATA_LOADING:
      return {
        ...state,
        isLoading: action.payload
      };
    case ActionType.SET_DATA_ERROR:
      return {
        ...state,
        isError: action.payload
      };
    case ActionType.SET_DATA:
      return {
        ...state,
        data: action.payload
      };
    case ActionType.SET_INPUT_VALUE:
      return {
        ...state,
        inputValue: action.payload,
        inputHasChanged: action.payload.length > 0
      };
    case ActionType.SET_SORTING:
      return {
        ...state,
        sorting: action.payload
      };
    default:
      return state;
  }
}

export const State = createContext<{
  state: AppState;
  dispatch: React.Dispatch<Action>;
}>({
  state: initialState,
  dispatch: () => null
});

export function StateProvider({ children }: PropsWithChildren) {
  const [state, dispatch] = useReducer(reducer, initialState);

  // Runs only once
  const { data: flightData, isLoading, isError } = useFlights();

  //  Update the app state when data is loaded
  useEffect(() => {
    dispatch({ type: ActionType.SET_DATA_LOADING, payload: isLoading });
    dispatch({ type: ActionType.SET_DATA_ERROR, payload: isError });
    dispatch({ type: ActionType.SET_DATA, payload: flightData });
  }, [flightData, isLoading, isError]);

  return <State.Provider value={{ state, dispatch }}>{children}</State.Provider>;
}
