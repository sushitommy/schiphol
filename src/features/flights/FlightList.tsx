import { State } from "@app/State";
import { useContext } from "react";
import { Flight } from "types";
import { sortFlightsByDateAndTime } from "utils/sortFlights";
import styles from "./FlightList.module.css";
import FlightListItem from "./FlightListItem";
import InfoMessage from "./InfoMessage";
import Sort from "./Sort";

function FlightList() {
  const { state } = useContext(State);

  if (!state.data) {
    return null;
  }

  // Check input
  if (!state.inputHasChanged) {
    return null;
  }

  if (state.inputValue.length < 3) {
    return <InfoMessage message="Please use 3 or more characters when searching." />;
  }

  let parsedFlights = state.data;

  // Filter on destination
  const airportRegex = new RegExp(state.inputValue, "i");
  parsedFlights = parsedFlights.filter((flight) => airportRegex.test(flight.airport));

  // Limit to 5 results
  parsedFlights = parsedFlights.slice(0, 5);

  // Sort flights
  if (parsedFlights.length > 1) {
    parsedFlights = sortFlightsByDateAndTime(parsedFlights, state.sorting);
  }

  if (parsedFlights.length > 0) {
    return (
      <div className={styles.flightList}>
        <Sort />
        {parsedFlights?.map((flight: Flight) => (
          <FlightListItem key={flight.flightIdentifier} {...flight} />
        ))}
      </div>
    );
  } else {
    return <div className={styles.flightList}>No flights found.</div>;
  }
}

export default FlightList;
