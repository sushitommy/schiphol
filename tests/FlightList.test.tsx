import { State } from "@app/State";
import * as useFlightsHookMethods from "@hooks/useFlights";
import { cleanup, render } from "@testing-library/react";
import { SortingType } from "types";
import { afterEach, describe, expect, test, vi } from "vitest";
import FlightList from "../src/features/flights/FlightList";
import flightListTestData from "./data/flightListTestData";

describe("FlightList", () => {
  const useFlightsSpy = vi.spyOn(useFlightsHookMethods, "default");
  useFlightsSpy.mockReturnValue({ data: flightListTestData, isLoading: false, isError: false });

  const dispatch = () => null;

  test("FlightList should render nothing when input is empty", () => {
    const state = {
      inputValue: "",
      inputHasChanged: false,
      data: flightListTestData,
      isLoading: false,
      isError: false,
      sorting: SortingType.EARLY_TO_LATE
    };

    const { asFragment } = render(
      <State.Provider value={{ state, dispatch }}>
        <FlightList />
      </State.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("FlightList should render 5 london items in order when input is 'Lon' and sorting is 'Early to late'", () => {
    const state = {
      inputValue: "lon",
      inputHasChanged: true,
      data: flightListTestData,
      isLoading: false,
      isError: false,
      sorting: SortingType.EARLY_TO_LATE
    };

    const { asFragment } = render(
      <State.Provider value={{ state, dispatch }}>
        <FlightList />
      </State.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("FlightList should render 5 london items in order when input is 'Lon' and sorting is 'Late to early'", () => {
    const state = {
      inputValue: "lon",
      inputHasChanged: true,
      data: flightListTestData,
      isLoading: false,
      isError: false,
      sorting: SortingType.LATE_TO_EARLY
    };

    const { asFragment } = render(
      <State.Provider value={{ state, dispatch }}>
        <FlightList />
      </State.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  afterEach(() => {
    useFlightsSpy.mockRestore();
    cleanup();
  });
});
