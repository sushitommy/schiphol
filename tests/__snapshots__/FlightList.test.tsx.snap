np// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FlightList > FlightList should render 5 london items in order when input is 'Lon' and sorting is 'Early to late' 1`] = `
<DocumentFragment>
  <div
    class="_flightList_36ec21"
  >
    <div
      class="_sortForm_f68e6c"
    >
      Sort by:
      <select
        class="_selectBox_f68e6c"
        name="sortOption"
      >
        <option
          value="earlyToLate"
        >
          Early to late
        </option>
        <option
          value="lateToEarly"
        >
          Late to early
        </option>
      </select>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        KL 1019 to London Heathrow
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-22
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          14:40
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          14:40
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          153.7487
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401KL1019/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        BA 2761 to London Gatwick
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-23
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          14:40
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          14:40
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          155.64577
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401BA2761/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        EJU 8874 to London Gatwick
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-23
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          15:05
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          15:05
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          154.88477
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401EJU8874/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        EZY 3004 to London Stansted
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-24
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          14:45
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          14:45
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          156.19374
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401EZY3004/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        BA 435 to London Heathrow
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-24
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          14:55
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          14:55
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          154.88477
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401BA435/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FlightList > FlightList should render 5 london items in order when input is 'Lon' and sorting is 'Late to early' 1`] = `
<DocumentFragment>
  <div
    class="_flightList_36ec21"
  >
    <div
      class="_sortForm_f68e6c"
    >
      Sort by:
      <select
        class="_selectBox_f68e6c"
        name="sortOption"
      >
        <option
          value="earlyToLate"
        >
          Early to late
        </option>
        <option
          value="lateToEarly"
        >
          Late to early
        </option>
      </select>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        BA 435 to London Heathrow
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-24
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          14:55
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          14:55
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          154.88477
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401BA435/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        EZY 3004 to London Stansted
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-24
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          14:45
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          14:45
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          156.19374
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401EZY3004/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        EJU 8874 to London Gatwick
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-23
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          15:05
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          15:05
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          154.88477
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401EJU8874/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        BA 2761 to London Gatwick
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-23
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          14:40
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          14:40
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          155.64577
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401BA2761/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
    <div
      class="_item_8283a2"
    >
      <h2
        class="_itemTitle_8283a2"
      >
        KL 1019 to London Heathrow
      </h2>
      <dl
        class="_itemContent_8283a2"
      >
        <dt>
          Date
        </dt>
        <dd>
          2022-02-22
        </dd>
        <dt>
          Original time
        </dt>
        <dd>
          14:40
        </dd>
        <dt>
          Expected time
        </dt>
        <dd>
          14:40
        </dd>
        <dt>
          Score
        </dt>
        <dd>
          153.7487
        </dd>
        <dt>
          <a
            href="/en/departures/flight/D20190401KL1019/"
          >
            Go to flight ↗
          </a>
        </dt>
        <dd />
      </dl>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FlightList > FlightList should render nothing when input is empty 1`] = `<DocumentFragment />`;
